File: numerouno/numerouno/page/habc2_examination_form/habc2_examination_form.js
frappe.provide("numerouno.habc2");

function h2Escape(value) {
	if (frappe.utils && typeof frappe.utils.escape_html === "function") {
		return frappe.utils.escape_html(value || "");
	}
	return $("<div>").text(value || "").html();
}

frappe.pages["habc2-examination-form"].on_page_load = function (wrapper) {
	if (wrapper.habc2_examination_form) {
		return;
	}
	try {
		if (!$("#habc2-examination-css").length) {
			$(
				'<link id="habc2-examination-css" rel="stylesheet" type="text/css" href="/assets/numerouno/css/habc2_examination_form.css?v=3">'
			).appendTo("head");
		}
		const page =
			wrapper.page && wrapper.page.main && wrapper.page.main.length
				? wrapper.page
				: frappe.ui.make_app_page({
						parent: wrapper,
						title: __("HABC2 Examination Declaration"),
						single_column: true,
				  });
		page.main.addClass("habc2-examination-page");
		wrapper.habc2_examination_form = new numerouno.habc2.Form(page);
	} catch (error) {
		console.error(error);
		$(wrapper)
			.find(".layout-main-section")
			.addBack(wrapper)
			.last()
			.html(
				`<div class="p-4">
					<h4>${__("HABC2 Examination Declaration")}</h4>
					<p class="text-danger">${__("Page failed to load. Press Ctrl+Shift+R and try again.")}</p>
				</div>`
			);
	}
};

frappe.pages["habc2-examination-form"].on_page_show = function (wrapper) {
	if (!wrapper.habc2_examination_form) {
		frappe.pages["habc2-examination-form"].on_page_load(wrapper);
		return;
	}
	wrapper.habc2_examination_form.resolve_route_and_load();
};

numerouno.habc2.Form = class {
	constructor(page) {
		this.page = page;
		this.doc = null;
		this.group = null;
		this.group_field = null;
		this.$root = $('<div class="h2-root"></div>').appendTo(this.page.main);
		this.bind_root_events();
		this.resolve_route_and_load();
	}

	bind_root_events() {
		this.$root.on("click", ".h2-load-group", (e) => {
			e.preventDefault();
			this.load_group();
		});
		this.$root.on("click", ".h2-save-btn", (e) => {
			e.preventDefault();
			this.save();
		});
		this.$root.on("click", ".h2-print-btn", (e) => {
			e.preventDefault();
			this.print_doc();
		});
		this.$root.on("click", ".h2-submit-btn", (e) => {
			e.preventDefault();
			this.submit_doc();
		});
		this.$root.on("click", ".h2-back-btn", (e) => {
			e.preventDefault();
			this.show_picker();
		});
		this.$root.on("click", ".h2-sign-clear", (e) => {
			e.preventDefault();
			this.clear_signature($(e.currentTarget).closest(".h2-sign-wrap"));
		});
		this.$root.on("keydown", ".h2-group-field-wrap input", (e) => {
			if (e.key === "Enter") {
				e.preventDefault();
				this.load_group();
			}
		});
	}

	resolve_route_and_load() {
		const route = frappe.get_route() || [];
		const docname = (route[1] || "").trim();
		const student_group = (frappe.route_options?.student_group || "").trim();
		if (frappe.route_options) {
			delete frappe.route_options.student_group;
		}
		if (docname) {
			this.fetch_form({ docname });
			return;
		}
		if (student_group) {
			this.fetch_form({ student_group });
			return;
		}
		if (this.doc && this.doc.name) {
			return;
		}
		this.show_picker(this.group);
	}

	show_picker(default_group = null) {
		this.doc = null;
		this.page.clear_inner_toolbar();
		this.page.set_title(__("HABC2 Examination Declaration"));
		this.page.set_primary_action(__("Open Form"), () => this.load_group());
		this.$root.html(`
			<div class="h2-portal-header">
				<div class="h2-kicker">${__("Highfield / HABC2")}</div>
				<h3>${__("Examination Declaration and Learner List")}</h3>
				<p>${__("Select a student group to fill the Highfield International form. Layout matches the official paper.")}</p>
			</div>
			<div class="h2-card">
				<div class="h2-group-picker">
					<div class="h2-group-field-wrap"></div>
					<button type="button" class="h2-btn h2-btn-primary h2-load-group">${__("Open Form")}</button>
				</div>
			</div>
		`);
		this.group_field = frappe.ui.form.make_control({
			df: {
				fieldtype: "Link",
				options: "Student Group",
				fieldname: "student_group",
				label: __("Student Group"),
				placeholder: __("Select Student Group"),
				reqd: 1,
			},
			parent: this.$root.find(".h2-group-field-wrap"),
			render_input: true,
		});
		this.group_field.make();
		this.group_field.refresh();
		this.$root.find(".h2-group-field-wrap .help-box").hide();
		if (default_group) {
			this.group_field.set_value(default_group);
		}
	}

	get_student_group() {
		return String(
			this.group_field?.get_value() ||
				this.group_field?.$input?.val() ||
				this.$root.find(".h2-group-field-wrap input").val() ||
				""
		).trim();
	}

	load_group() {
		const student_group = this.get_student_group();
		if (!student_group) {
			frappe.msgprint(__("Please select a Student Group"));
			return;
		}
		this.fetch_form({ student_group });
	}

	fetch_form(args) {
		frappe.call({
			method: "numerouno.numerouno.page.habc2_examination_form.habc2_examination_form.get_form_html",
			args,
			freeze: true,
			freeze_message: __("Loading Highfield form..."),
			callback: (r) => {
				if (r.exc) {
					return;
				}
				this.doc = r.message.doc;
				this.group = this.doc.student_group;
				this.render(r.message.html);
				if (this.doc.name) {
					frappe.router.replace_route("habc2-examination-form", this.doc.name);
				}
			},
		});
	}

	render(html) {
		this.page.clear_primary_action();
		this.page.set_title(this.doc.name || __("HABC2 Examination Declaration"));
		this.page.set_primary_action(__("Save"), () => this.save());
		this.page.set_secondary_action(__("Print"), () => this.print_doc());
		const submitted = h2Cint(this.doc.docstatus) === 1;
		this.$root.html(`
			<div class="h2-form-bar">
				<button type="button" class="h2-btn h2-back-btn">${__("Change Group")}</button>
				<div class="h2-form-bar-meta">
					<strong>${h2Escape(this.doc.qualification_unit_title || this.doc.student_group || "")}</strong>
					<span>${h2Escape(this.doc.name || "")}</span>
				</div>
				<button type="button" class="h2-btn h2-print-btn">${__("Print")}</button>
				<button type="button" class="h2-btn h2-submit-btn" ${submitted ? "disabled" : ""}>${__("Submit")}</button>
				<button type="button" class="h2-btn h2-btn-primary h2-save-btn" ${submitted ? "disabled" : ""}>${__("Save")}</button>
			</div>
			<div class="h2-paper-wrap ${submitted ? "is-readonly" : ""}">${html}</div>
		`);
		this.init_signature_canvases();
		if (submitted) {
			this.$root.find("input, textarea, button.h2-sign-clear").prop("disabled", true);
		}
	}

	collect() {
		const data = Object.assign({}, this.doc);
		this.$root.find(".h2-input[data-field]").each((_, el) => {
			const $el = $(el);
			data[$el.data("field")] = $el.val();
		});
		const bySr = {};
		(this.doc.learners || []).forEach((row) => {
			bySr[String(row.sr_no)] = Object.assign({}, row);
		});
		this.$root.find("[data-row-field]").each((_, el) => {
			const $el = $(el);
			const sr = String($el.data("sr"));
			if (!bySr[sr]) {
				bySr[sr] = { sr_no: h2Cint(sr) };
			}
			const field = $el.data("row-field");
			if ($el.attr("type") === "checkbox") {
				bySr[sr][field] = $el.is(":checked") ? 1 : 0;
			} else if ($el.attr("type") === "radio") {
				if ($el.is(":checked")) {
					bySr[sr][field] = $el.val();
				}
			} else {
				bySr[sr][field] = $el.val();
			}
		});
		data.learners = Object.keys(bySr)
			.sort((a, b) => h2Cint(a) - h2Cint(b))
			.map((key) => bySr[key]);
		return data;
	}

	save() {
		if (h2Cint(this.doc.docstatus) === 1) {
			frappe.msgprint(__("Submitted form cannot be edited"));
			return;
		}
		frappe.call({
			method: "numerouno.numerouno.page.habc2_examination_form.habc2_examination_form.save_form_data",
			args: { data: this.collect() },
			freeze: true,
			freeze_message: __("Saving..."),
			callback: (r) => {
				if (r.exc) {
					return;
				}
				this.doc = r.message;
				frappe.show_alert({ message: __("Saved {0}", [this.doc.name]), indicator: "green" });
				frappe.router.replace_route("habc2-examination-form", this.doc.name);
			},
		});
	}

	submit_doc() {
		if (!this.doc || !this.doc.name) {
			this.save();
			return;
		}
		frappe.confirm(__("Submit this Highfield declaration?"), () => {
			this.save_then(() => {
				frappe.call({
					method: "numerouno.numerouno.page.habc2_examination_form.habc2_examination_form.submit_form",
					args: { docname: this.doc.name },
					freeze: true,
					callback: (r) => {
						if (r.exc) {
							return;
						}
						this.fetch_form({ docname: this.doc.name });
					},
				});
			});
		});
	}

	save_then(fn) {
		frappe.call({
			method: "numerouno.numerouno.page.habc2_examination_form.habc2_examination_form.save_form_data",
			args: { data: this.collect() },
			freeze: true,
			callback: (r) => {
				if (r.exc) {
					return;
				}
				this.doc = r.message;
				fn();
			},
		});
	}

	print_doc() {
		if (!this.doc || !this.doc.name) {
			frappe.msgprint(__("Save the form before printing"));
			return;
		}
		this.save_then(() => {
			window.open(
				frappe.urllib.get_full_url(
					"/printview?doctype=" +
						encodeURIComponent("HABC2 Examination Declaration") +
						"&name=" +
						encodeURIComponent(this.doc.name) +
						"&format=" +
						encodeURIComponent("HABC2 Examination Declaration") +
						"&no_letterhead=1"
				),
				"_blank"

			);
		});
	}

	init_signature_canvases() {
		this.$root.find(".h2-sign-canvas").each((_, canvas) => {
			this.bind_signature_canvas(canvas);
		});
	}

	bind_signature_canvas(canvas) {
		if (!canvas || canvas.dataset.bound === "1") {
			return;
		}
		canvas.dataset.bound = "1";
		const ctx = canvas.getContext("2d");
		ctx.strokeStyle = "#111";
		ctx.lineWidth = 2;
		ctx.lineCap = "round";
		let drawing = false;
		const coords = (e) => {
			const rect = canvas.getBoundingClientRect();
			const src = e.touches && e.touches[0] ? e.touches[0] : e;
			return {
				x: (src.clientX - rect.left) * (canvas.width / rect.width),
				y: (src.clientY - rect.top) * (canvas.height / rect.height),
			};
		};
		const start = (e) => {
			e.preventDefault();
			drawing = true;
			const p = coords(e);
			ctx.beginPath();
			ctx.moveTo(p.x, p.y);
		};
		const move = (e) => {
			if (!drawing) {
				return;
			}
			e.preventDefault();
			const p = coords(e);
			ctx.lineTo(p.x, p.y);
			ctx.stroke();
		};
		const end = () => {
			if (!drawing) {
				return;
			}
			drawing = false;
			const $wrap = $(canvas).closest(".h2-sign-wrap");
			$wrap.find("input[data-field], input[data-row-field]").val(canvas.toDataURL("image/png"));
			$wrap.find("img").remove();
		};
		canvas.addEventListener("mousedown", start);
		canvas.addEventListener("mousemove", move);
		window.addEventListener("mouseup", end);
		canvas.addEventListener("touchstart", start, { passive: false });
		canvas.addEventListener("touchmove", move, { passive: false });
		canvas.addEventListener("touchend", end);
	}

	clear_signature($wrap) {
		const canvas = $wrap.find("canvas").get(0);
		if (canvas) {
			const ctx = canvas.getContext("2d");
			ctx.clearRect(0, 0, canvas.width, canvas.height);
		}
		$wrap.find("img").remove();
		$wrap.find("input[data-field], input[data-row-field]").val("");
	}
};

function h2Cint(value) {
	if (typeof window.cint === "function") {
		return window.cint(value);
	}
	const parsed = parseInt(value, 10);
	return Number.isNaN(parsed) ? 0 : parsed;
}

File: numerouno/numerouno/page/habc2_examination_form/habc2_examination_form.py
import frappe
from frappe.utils import cint, getdate, today

from numerouno.numerouno.doctype.habc2_examination_declaration.habc2_examination_declaration import (
	DEFAULT_CENTRE_NO,
	_split_name,
	group_defaults,
	paginate_learners,
)
from numerouno.numerouno.utils.signatures import is_empty_signature, resolve_learner_signature

DOCTYPE = "HABC2 Examination Declaration"
TEMPLATE = "numerouno/numerouno/print_format/habc2_examination_declaration/habc2_examination_declaration.html"

PARENT_FIELDS = (
	"naming_series",
	"student_group",
	"course",
	"centre_name",
	"centre_number",
	"nominated_tutor",
	"nominated_tutor_name",
	"tutor_number",
	"examination_venue",
	"course_exam_id",
	"number_of_learners",
	"qualification_unit_title",
	"examination_date",
	"exam_start_time",
	"exam_end_time",
	"course_from",
	"course_to",
	"invigilator_name",
	"invigilator_position",
	"invigilator_signature",
	"invigilator_date",
	"comments",
)


def _as_date(value):
	if not value:
		return ""
	try:
		return str(getdate(value))
	except Exception:
		return str(value)


def _as_time(value):
	if value in (None, ""):
		return ""
	text = str(value)
	if ":" in text:
		parts = text.split(":")
		return f"{parts[0].zfill(2)}:{parts[1].zfill(2)}"
	return text


def _serialize_row(row, student_group=None):
	signature = resolve_learner_signature(row.get("student"), student_group, row.get("signature"))
	names = row.get("learner_names") or ""
	surname = row.get("learner_surname") or ""
	if row.get("student"):
		names, surname = _split_name(row.get("student"))
	return {
		"sr_no": row.get("sr_no"),
		"student": row.get("student") or "",
		"learner_names": names,
		"learner_surname": surname,
		"signature": signature or "",
		"date_of_birth": _as_date(row.get("date_of_birth")),
		"id_type": row.get("id_type") or "",
		"telephone": row.get("telephone") or "",
		"retake": cint(row.get("retake")),
		"gender": row.get("gender") or "",
	}


def _serialize_doc(doc):
	_ensure_learner_signatures(doc)
	_ensure_learner_names(doc)
	learners = [_serialize_row(row, doc.student_group) for row in (doc.learners or [])]
	pages = paginate_learners(learners)
	data = {field: doc.get(field) for field in PARENT_FIELDS}
	data.update(
		{
			"name": doc.name,
			"docstatus": doc.docstatus,
			"examination_date": _as_date(doc.examination_date),
			"invigilator_date": _as_date(doc.invigilator_date),
			"course_from": _as_date(doc.course_from),
			"course_to": _as_date(doc.course_to),
			"exam_start_time": _as_time(doc.exam_start_time),
			"exam_end_time": _as_time(doc.exam_end_time),
			"learners": learners,
			"learner_pages": pages,
		}
	)
	return data


def _ensure_learner_names(doc):
	if cint(doc.docstatus) != 0:
		return
	for row in doc.learners or []:
		if not row.get("student"):
			continue
		names, surname = _split_name(row.get("student"))
		if (row.get("learner_names") or "") == names and (row.get("learner_surname") or "") == surname:
			continue
		row.learner_names = names
		row.learner_surname = surname
		if row.get("name"):
			frappe.db.set_value(
				"HABC2 Learner",
				row.name,
				{"learner_names": names, "learner_surname": surname},
				update_modified=False,
			)


def _ensure_learner_signatures(doc):
	changed = False
	for row in doc.learners or []:
		if not row.get("student"):
			continue
		sig = resolve_learner_signature(row.get("student"), doc.student_group, row.get("signature"))
		if sig and is_empty_signature(row.get("signature")):
			row.signature = sig
			changed = True
			if row.get("name") and cint(doc.docstatus) == 0:
				frappe.db.set_value("HABC2 Learner", row.name, "signature", sig, update_modified=False)
	return changed


def _apply_payload(doc, data):
	for field in PARENT_FIELDS:
		if field in data:
			doc.set(field, data.get(field))
	doc.learners = []
	for row in data.get("learners") or []:
		if not (
			row.get("student")
			or row.get("learner_names")
			or row.get("learner_surname")
			or row.get("signature")
			or row.get("telephone")
			or row.get("id_type")
		):
			continue
		doc.append(
			"learners",
			{
				"sr_no": row.get("sr_no"),
				"student": row.get("student") or None,
				"learner_names": row.get("learner_names") or "",
				"learner_surname": row.get("learner_surname") or "",
				"signature": row.get("signature") or "",
				"date_of_birth": row.get("date_of_birth") or None,
				"id_type": row.get("id_type") or "",
				"telephone": row.get("telephone") or "",
				"retake": cint(row.get("retake")),
				"gender": row.get("gender") or "",
			},
		)


def _get_or_create(student_group):
	student_group = (student_group or "").strip()
	if not student_group:
		frappe.throw("Student Group is required")

	existing = frappe.db.get_value(
		DOCTYPE,
		{"student_group": student_group, "docstatus": ["<", 2]},
		"name",
	)
	if existing:
		return _fill_empty_draft(frappe.get_doc(DOCTYPE, existing), student_group)

	defaults = group_defaults(student_group)
	doc = frappe.new_doc(DOCTYPE)
	for field, value in defaults.items():
		if field == "learners":
			continue
		doc.set(field, value)
	if not doc.examination_date:
		doc.examination_date = today()
	if not doc.invigilator_date:
		doc.invigilator_date = today()
	for row in defaults.get("learners") or []:
		doc.append("learners", row)
	doc.insert(ignore_permissions=True)
	return doc


def _fill_empty_draft(doc, student_group):
	if cint(doc.docstatus) != 0:
		return doc
	has_learners = any(
		(row.student or row.learner_names or row.learner_surname) for row in (doc.learners or [])
	)
	if has_learners:
		return doc
	defaults = group_defaults(student_group)
	for field, value in defaults.items():
		if field == "learners":
			continue
		if value and not doc.get(field):
			doc.set(field, value)
	doc.learners = []
	for row in defaults.get("learners") or []:
		doc.append("learners", row)
	doc.save(ignore_permissions=True)
	return doc


@frappe.whitelist()
def get_form_data(docname=None, student_group=None):
	docname = (docname or "").strip()
	student_group = (student_group or "").strip() or None
	if docname:
		doc = frappe.get_doc(DOCTYPE, docname)
	elif student_group:
		doc = _get_or_create(student_group)
	else:
		frappe.throw("Select a student group to open the HABC2 form.")
	if not doc.centre_number:
		doc.centre_number = DEFAULT_CENTRE_NO
		if doc.name and cint(doc.docstatus) == 0:
			doc.db_set("centre_number", DEFAULT_CENTRE_NO, update_modified=False)
	return _serialize_doc(doc)


def _doc_for_template(data):
	doc = frappe._dict(data)
	pages = []
	for page in data.get("learner_pages") or []:
		pages.append([frappe._dict(row) for row in page])
	doc.learner_pages = pages
	doc.learners = [frappe._dict(row) for row in data.get("learners") or []]
	doc.get_formatted = lambda field: doc.get(field) or ""
	return doc


@frappe.whitelist()
def get_form_html(docname=None, student_group=None):
	doc_data = get_form_data(docname=docname, student_group=student_group)
	html = frappe.render_template(TEMPLATE, {"doc": _doc_for_template(doc_data), "editable": 1})
	return {"doc": doc_data, "html": html}


@frappe.whitelist()
def save_form_data(data):
	data = frappe.parse_json(data)
	docname = (data.get("name") or "").strip()
	if docname:
		doc = frappe.get_doc(DOCTYPE, docname)
		if doc.docstatus == 1:
			frappe.throw("Submitted form cannot be edited")
	else:
		doc = frappe.new_doc(DOCTYPE)
	_apply_payload(doc, data)
	doc.save()
	return _serialize_doc(doc)


@frappe.whitelist()
def submit_form(docname):
	doc = frappe.get_doc(DOCTYPE, docname)
	if doc.docstatus == 0:
		doc.submit()
	return {"name": doc.name, "docstatus": doc.docstatus}


@frappe.whitelist()
def cancel_form(docname):
	doc = frappe.get_doc(DOCTYPE, docname)
	if doc.docstatus == 2:
		return {"name": doc.name, "docstatus": doc.docstatus}
	if doc.docstatus != 1:
		frappe.throw("Only a submitted HABC2 Examination Declaration can be cancelled")
	doc.cancel()
	return {"name": doc.name, "docstatus": doc.docstatus}


@frappe.whitelist()
def amend_form(docname):
	doc = frappe.get_doc(DOCTYPE, docname)
	if doc.docstatus != 2:
		frappe.throw("Only cancelled HABC2 Examination Declarations can be amended")
	amended = frappe.copy_doc(doc)
	amended.amended_from = doc.name
	amended.insert()
	return _serialize_doc(amended)
